source/common/ssl/ssl_socket.cc

```diff
@@ -153,57 +153,48 @@ Network::IoResult SslSocket::doWrite(Buffer::Instance& write_buffer, bool end_st
     }
   }
 
-  uint64_t original_buffer_length = write_buffer.length();
+  uint64_t bytes_to_write;
+  if (bytes_to_retry_) {
+    bytes_to_write = bytes_to_retry_;
+    bytes_to_retry_ = 0;
+  } else {
+    bytes_to_write = std::min(write_buffer.length(), static_cast<uint64_t>(16384));
+  }
+
   uint64_t total_bytes_written = 0;
-  bool keep_writing = true;
-  while ((original_buffer_length != total_bytes_written) && keep_writing) {
-    // Protect against stack overflow if the buffer has a very large buffer chain.
-    // TODO(mattklein123): See the comment on getRawSlices() for why we have to also check
-    // original_buffer_length != total_bytes_written during loop iteration.
+  while (bytes_to_write) {
     // TODO(mattklein123): As it relates to our fairness efforts, we might want to limit the number
     // of iterations of this loop, either by pure iterations, bytes written, etc.
-    const uint64_t MAX_SLICES = 32;
-    Buffer::RawSlice slices[MAX_SLICES];
-    uint64_t num_slices = write_buffer.getRawSlices(slices, MAX_SLICES);
-
-    uint64_t inner_bytes_written = 0;
-    for (uint64_t i = 0; (i < num_slices) && (original_buffer_length != total_bytes_written); i++) {
-      // SSL_write() requires that if a previous call returns SSL_ERROR_WANT_WRITE, we need to call
-      // it again with the same parameters. Most implementations keep track of the last write size.
-      // In our case we don't need to do that because: a) SSL_write() will not write partial
-      // buffers. b) We only move() into the write buffer, which means that it's impossible for a
-      // particular chain to increase in size. So as long as we start writing where we left off we
-      // are guaranteed to call SSL_write() with the same parameters.
-      int rc = SSL_write(ssl_.get(), slices[i].mem_, slices[i].len_);
-      ENVOY_CONN_LOG(trace, "ssl write returns: {}", callbacks_->connection(), rc);
-      if (rc > 0) {
-        inner_bytes_written += rc;
-        total_bytes_written += rc;
-      } else {
-        int err = SSL_get_error(ssl_.get(), rc);
-        switch (err) {
-        case SSL_ERROR_WANT_WRITE:
-          keep_writing = false;
-          break;
-        case SSL_ERROR_WANT_READ:
-        // Renegotiation has started. We don't handle renegotiation so just fall through.
-        default:
-          drainErrorQueue();
-          return {PostIoAction::Close, total_bytes_written, false};
-        }
 
+    // SSL_write() requires that if a previous call returns SSL_ERROR_WANT_WRITE, we need to call
+    // it again with the same parameters. This is done by tracking last write size, but not write
+    // data, since linearize() will return the same undrained data anyway.
+    ASSERT(bytes_to_write <= write_buffer.length());
+    int rc = SSL_write(ssl_.get(), write_buffer.linearize(bytes_to_write), bytes_to_write);
```

Comment:
stupid openssl/boringssl question: I don't suppose there is some writev equivalent? Seems a shame we need to linearize here.

Reply:
I don't think these is one?  Can we tag off-project people?  Maybe @davidben knows

Reply:
Sanity check: if we linearize multiple times it only does the work one time, correct?  I want to make sure there's no corner case where trickling bites gets super expensive.

Reply:
Unfortunately boringssl doesn't have a built in writev equivalent. You could potentially create a fake FD to use writev with and then dump that into SSL_write, but that probably has more overhead/complexity than linearize.

Reply:
We do have plans to add an in-place API so you don't need to copy things in and out of OpenSSL's goofy fd abstraction, though probably the first iteration will still require it be linearized. writev is a little tricky with block ciphers. (Though that work will make it easier to add something like writev later.)

Reply:
@mattklein123 Done.

@alyssawilk Yes, linearization will happen only once (for the same `size` argument).